### PR TITLE
[Feature] Add basic interpolater

### PR DESCRIPTION
**Pull request by Yshuo-Li**

### CODE REVIEW

File: mmedit/models/video_interpolater/basic_interpolater.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+        """Init weights for models.
+
+        Args:
+            pretrained (str, optional): Path for pretrained weights. If given
+                None, pretrained weights will not be loaded. Defaults to None.
+        """
+        self.generator.init_weights(pretrained)
+
+    @auto_fp16(apply_to=('inputs', ))
+    def forward(self, inputs, target=None, test_mode=False, **kwargs):
```
Comment (ckkelvinchan) [inline mmedit/models/video_interpolater/basic_interpolater.py:64]: Maybe we can use `input` here. Just like in video SR, we also do not use `lqs` and `gts`. We can simply use singular form here.

**Review by ckkelvinchan — COMMENTED** [review]
You can also consider inheriting from `BasicRestorer`, as many functions are identical.

File: mmedit/models/video_interpolater/basic_interpolater.py (hunk 1 above)
Comment (Yshuo-Li) [reply mmedit/models/video_interpolater/basic_interpolater.py:64]: `input` conflicts with Python's own functions `input()`

Comment (Yshuo-Li) [reply mmedit/models/video_interpolater/basic_interpolater.py:64]: `inputs` will highlight the difference between input and output

Comment (ckkelvinchan) [reply mmedit/models/video_interpolater/basic_interpolater.py:64]: Oops yes. Then we can just keep it as it is.

File: mmedit/models/video_interpolater/basic_interpolater.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+    def _save_image(meta, iteration, save_path, pred):
+        if 'target_path' in meta[0]:
+            pred_path = meta[0]['target_path']
+            folder_name = osp.splitext(osp.basename(pred_path))[0]
+        else:
+            pred_path1 = osp.splitext(osp.basename(
+                meta[0]['inputs_path'][0]))[0]
+            pred_path2 = osp.splitext(osp.basename(
+                meta[0]['inputs_path'][1]))[0]
+            folder_name = f'{pred_path1}_{pred_path2}'
```
Comment (ckkelvinchan) [inline mmedit/models/video_interpolater/basic_interpolater.py:168]: Here I am a bit confused. What are the formats of `target_path` and `inputs_path`? Could you provide an example?

File: mmedit/models/video_interpolater/basic_interpolater.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+
+        Args:
+            data_batch (dict): A batch of data.
+            kwargs (dict): Other arguments for ``val_step``.
+
+        Returns:
+            dict: Returned output.
+        """
+        output = self.forward_test(**data_batch, **kwargs)
+        return output
```
Comment (ckkelvinchan) [inline mmedit/models/video_interpolater/basic_interpolater.py:223]: Not covered by unittest.

File: mmedit/models/video_interpolater/basic_interpolater.py [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+from ..builder import build_backbone, build_loss
+from ..registry import MODELS
+
+
+@MODELS.register_module()
+class BasicInterpolater(BaseModel):
+    """Basic model for video interpolation.
+
+    It must contain a generator that takes frames as inputs and outputs a
+    interpolated frame. It also has a pixel-wise loss for training.
```
Comment (ckkelvinchan) [inline mmedit/models/video_interpolater/basic_interpolater.py:19]: outputs an interpolated frame.

File: mmedit/models/video_interpolater/basic_interpolater.py [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+        """Evaluation function.
+
+        Args:
+            output (Tensor): Model output with shape (n, c, h, w).
+            target (Tensor): GT Tensor with shape (n, c, h, w).
+
+        Returns:
+            dict: Evaluation results.
+        """
+        crop_border = self.test_cfg.crop_border
```
Comment (ckkelvinchan) [inline mmedit/models/video_interpolater/basic_interpolater.py:110]: `crop_border = self.test_cfg.get(crop_border, 0)` is better.

File: mmedit/models/video_interpolater/basic_interpolater.py [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+                     inputs,
+                     target=None,
+                     meta=None,
+                     save_image=False,
+                     save_path=None,
+                     iteration=None):
+        """Testing forward function.
+
+        Args:
+            inputs (Tensor): Tensor of inputs frames with shape (n, c, h, w).
```
Comment (ckkelvinchan) [inline mmedit/models/video_interpolater/basic_interpolater.py:131]: Should be `(n, 2, c, h, w)`?

File: mmedit/models/video_interpolater/basic_interpolater.py [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+            num_samples=len(target.data),
+            results=dict(
+                inputs=inputs.cpu(), target=target.cpu(), output=output.cpu()))
+        return outputs
+
+    def evaluate(self, output, target):
+        """Evaluation function.
+
+        Args:
+            output (Tensor): Model output with shape (n, c, h, w).
```
Comment (ckkelvinchan) [inline mmedit/models/video_interpolater/basic_interpolater.py:104]: (n, 2, c, h, w)

File: mmedit/models/video_interpolater/basic_interpolater.py [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+        if test_mode:
+            return self.forward_test(inputs, target, **kwargs)
+
+        return self.forward_train(inputs, target)
+
+    def forward_train(self, inputs, target):
+        """Training forward function.
+
+        Args:
+            inputs (Tensor): Tensor of inputs frames with shape (n, c, h, w).
```
Comment (ckkelvinchan) [inline mmedit/models/video_interpolater/basic_interpolater.py:83]: (n, 2, c, h, w)

File: mmedit/models/video_interpolater/basic_interpolater.py (hunk 7 above)
Comment (Yshuo-Li) [reply mmedit/models/video_interpolater/basic_interpolater.py:104]: The output is shaped in (n, 2, c, h, w)

File: mmedit/models/video_interpolater/basic_interpolater.py [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+        loss_pix = self.pixel_loss(output, target)
+        losses['loss_pix'] = loss_pix
+        outputs = dict(
+            losses=losses,
+            num_samples=len(target.data),
+            results=dict(
+                inputs=inputs.cpu(), target=target.cpu(), output=output.cpu()))
+        return outputs
+
+    def evaluate(self, output, target):
```
Comment (ckkelvinchan) [inline mmedit/models/video_interpolater/basic_interpolater.py:105]: If this function is for a model that produces outputs with `(n, c, h, w)`, we do not need to consider the case of  `(n, t, c, h, w)`.

File: mmedit/models/video_interpolater/basic_interpolater.py [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+            meta (list[dict]): Meta data, such as path of target file.
+                Default: None. These dictionaries should contain
+                'target_path' (str of a path) or 'inputs_path' (list of str)
+            iteration (int): Iteration for the saving image name.
+                Default: None.
+            save_path (str): Path to save image. Default: None.
+            output (Tensor): Output image.
+        """
+
+        if output.ndim == 4:  # an image, key = 000001/0000 (Vimeo-90K)
```
Comment (ckkelvinchan) [inline mmedit/models/video_interpolater/basic_interpolater.py:196]: Similar here. The outputs must be of shape `(n c, h, w)`.

File: mmedit/models/video_interpolater/basic_interpolater.py [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+                else:
+                    raise ValueError('iteration should be number or None, '
+                                     f'but got {type(iteration)}')
+                mmcv.imwrite(tensor2img(output[:, i, :, :, :]), save_path_i)
+
+    def forward_dummy(self, img):
+        """Used for computing network FLOPs.
+
+        Args:
+            img (Tensor): Input image.
```
Comment (ckkelvinchan) [inline mmedit/models/video_interpolater/basic_interpolater.py:225]: images

File: mmedit/models/video_interpolater/basic_interpolater.py [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+            kwargs (dict): Other arguments for ``val_step``.
+
+        Returns:
+            dict: Returned output.
+        """
+        output = self.forward_test(**data_batch, **kwargs)
+        return output
+
+    @staticmethod
+    def generate_frames(input_images, output_images):
```
Comment (ckkelvinchan) [inline mmedit/models/video_interpolater/basic_interpolater.py:268]: This function will be used in the demo right?

By the way, how does the demo split a video into multiple inputs?

File: mmedit/models/video_interpolater/basic_interpolater.py (hunk 9 above)
Comment (Yshuo-Li) [reply mmedit/models/video_interpolater/basic_interpolater.py:105]: The output of CAIN is an image, but maybe some methods produce more than one image.

File: mmedit/models/video_interpolater/basic_interpolater.py [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
+                if isinstance(iteration, numbers.Number):
+                    save_path_i = osp.join(save_path, folder_name,
+                                           f'{i:08d}-{iteration + 1:06d}.png')
+                elif iteration is None:
+                    save_path_i = osp.join(save_path, folder_name,
+                                           f'{i:08d}.png')
+                else:
+                    raise ValueError('iteration should be number or None, '
+                                     f'but got {type(iteration)}')
+                mmcv.imwrite(tensor2img(output[:, i, :, :, :]), save_path_i)
```
[the remark this answers is not in the document]
Comment (Yshuo-Li) [reply mmedit/models/video_interpolater/basic_interpolater.py:226]: This function is mainly used to save valid results.

File: mmedit/models/video_interpolater/basic_interpolater.py [hunk 14]
```diff
@@ -0,0 +1,296 @@
...
+                 num_inputs=2,
+                 step=1,
+                 train_cfg=None,
+                 test_cfg=None,
+                 pretrained=None):
+        super().__init__()
+
+        self.train_cfg = train_cfg
+        self.test_cfg = test_cfg
+        self.num_inputs = num_inputs
```
[the remark this answers is not in the document]
Comment (Yshuo-Li) [reply mmedit/models/video_interpolater/basic_interpolater.py:49]: in the demo.
Interpolation demo need the `num_inputs` and `step`

File: mmedit/models/video_interpolater/basic_interpolater.py [hunk 15]
```diff
@@ -0,0 +1,308 @@
...
+
+        Returns:
+            list[Tensor]: Splided tensors.
+        """
+
+        num_frames = input_tensors.shape[0]
+
+        result = []
+        for i in range(0, num_frames - 1, 1):
+            result.append(input_tensors[i:i + 2])
```
Comment (ckkelvinchan) [inline mmedit/models/video_interpolater/basic_interpolater.py:282]: Suggestions:
1. We can follow your previous approach of using `num_input_frames` as one argument. In this way, CAIN (2 frames->1 frame) and QVI (4 frames->1 frame) can both use this class. The argument `step` can be omitted.
2. We don't need to make it a static method. In such a way, `self.num_input_frames` can be passed to this function, if the first point is considered.

File: mmedit/models/video_interpolater/basic_interpolater.py [hunk 16]
```diff
@@ -0,0 +1,308 @@
...
+                if isinstance(iteration, numbers.Number):
+                    save_path_i = osp.join(save_path, folder_name,
+                                           f'{i:08d}-{iteration + 1:06d}.png')
+                elif iteration is None:
+                    save_path_i = osp.join(save_path, folder_name,
+                                           f'{i:08d}.png')
+                else:
+                    raise ValueError('iteration should be number or None, '
+                                     f'but got {type(iteration)}')
+                mmcv.imwrite(tensor2img(output[:, i, :, :, :]), save_path_i)
```
Comment (ckkelvinchan) [inline mmedit/models/video_interpolater/basic_interpolater.py:216]: As discussed:

1. If the output has only 4 dimensions, then the convention is the middle frame. Then we can use `00000000_050.png`.

2. If the output has 5 dimensions, we assume they are multiple intermediate frames between 2 frames. For example, in CAIN, the output would be `torch.stack([x0.25, 0.5, 0.75], dim=1)`. Then we can use something like `00000000_025.png`, `00000000_050.png`, `00000000_075.png`, assuming x0 has name `00000000.png`.

File: mmedit/models/video_interpolater/basic_interpolater.py [hunk 17]
```diff
@@ -0,0 +1,308 @@
...
+                     save_path=None,
+                     iteration=None):
+        """Testing forward function.
+
+        This is a basic function, interpolate a frame between the given two
+        frames.
+
+        Args:
+            inputs (Tensor): Tensor of inputs frames with shape
+                (n, 2, c, h, w).
```
Comment (ckkelvinchan) [inline mmedit/models/video_interpolater/basic_interpolater.py:150]: Tensor of input frames

(Do not specify shape)

File: mmedit/models/video_interpolater/basic_interpolater.py [hunk 18]
```diff
@@ -0,0 +1,308 @@
...
+            target (Tensor): Tensor of target frame with shape (n, c, h, w).
```
Comment (ckkelvinchan) [inline mmedit/models/video_interpolater/basic_interpolater.py:151]: Tensor of target frame(s)

File: mmedit/models/video_interpolater/basic_interpolater.py [hunk 19]
```diff
@@ -0,0 +1,308 @@
...
+        Returns:
+            list[np.array]: The final frames.
+                in_frame, out_frame, in_frame, out_frame, in_frame ...
+        """
+
+        result = []
+        for i in range(len(output_images)):
+            result.append(input_images[i])
+            result.append(output_images[i])
+        result.append(input_images[-1])
```
Comment (ckkelvinchan) [inline mmedit/models/video_interpolater/basic_interpolater.py:306]: Sorry, I just find one use case of using Tensor in this function. Consider the case when a model takes two frames as input and interpolates 3 frames (t=0.25, 0.5, 0.75). In this case, the `input_images` should be 
[x0, x1, x2] and `output_images` is [y0, y1], where 
```
y0 = torch.stack([x0.25, x0.5, x0.75], dim=1)
y1 = torch.stack([x1.25, x1.5, x1.75], dim=1)
```
It would be easier if `input_images` and `output_images` are Tensors and we transform them to image here:
```
result = []
for i, output in enumerate(output_images):
    result.append(tensor2img(input_images[i]))
    for j in range(0, output.size(1)):
        result.append(tensor2img(output[:, j, :, :, :])
result.append(input_images[-1])

return result
```

File: mmedit/models/video_interpolater/basic_interpolater.py [hunk 20]
```diff
@@ -0,0 +1,307 @@
...
+        return self.forward_train(inputs, target)
+
+    def forward_train(self, inputs, target):
+        """Training forward function.
+
+        This is a basic function, interpolate a frame between the given two
+        frames.
+
+        Args:
+            inputs (Tensor): Tensor of inputs frame(s).
```
Comment (ckkelvinchan) [inline mmedit/models/video_interpolater/basic_interpolater.py:87]: input frames (not inputs frames)

File: mmedit/models/video_interpolater/basic_interpolater.py [hunk 21]
```diff
@@ -0,0 +1,307 @@
...
+        """split input tensors for inference.
+
+        This is a basic function, interpolate a frame between the given two
+        frames.
+
+        Args:
+            input_tensors (Tensor): Tensor of input frames.
+
+        Returns:
+            list[Tensor]: Splided tensors.
```
Comment (ckkelvinchan) [inline mmedit/models/video_interpolater/basic_interpolater.py:274]: A list of split tensors

File: mmedit/models/video_interpolater/basic_interpolater.py [hunk 22]
```diff
@@ -0,0 +1,307 @@
...
+        Returns:
+            list[np.array]: The final frames.
+                in_frame, out_frame, in_frame, out_frame, in_frame ...
+        """
+
+        result = []
+        for i in range(len(input_tensors)):
+            result.append(tensor2img(input_tensors[i][:1]))
+            result.append(tensor2img(output_tensors[i][:1]))
+        result.append(tensor2img(input_tensors[-1][1:]))
```
Comment (ckkelvinchan) [inline mmedit/models/video_interpolater/basic_interpolater.py:305]: input_tensors[i][.squeeze(0)

File: mmedit/models/video_interpolater/basic_interpolater.py [hunk 23]
```diff
@@ -0,0 +1,307 @@
...
+                None, pretrained weights will not be loaded. Defaults to None.
+        """
+        self.generator.init_weights(pretrained)
+
+    @auto_fp16(apply_to=('inputs', ))
+    def forward(self, inputs, target=None, test_mode=False, **kwargs):
+        """Forward function.
+
+        Args:
+            inputs (Tensor): Tensor of inputs frames.
```
Comment (ckkelvinchan) [inline mmedit/models/video_interpolater/basic_interpolater.py:69]: input frames

File: mmedit/models/video_interpolater/basic_interpolater.py (hunk 22 above)
Comment (Yshuo-Li) [reply mmedit/models/video_interpolater/basic_interpolater.py]: It is not the same.

**Review by ckkelvinchan — APPROVED** [review]
Remember to check for "merge" and "split" when writing the demo.
